### PR TITLE
FEAT: optimize pyav

### PR DESCRIPTION
**Pull request by FirefoxMetzger**
Closes: https://github.com/imageio/imageio/issues/926, https://github.com/imageio/imageio/issues/929, https://github.com/imageio/imageio/issues/932, https://github.com/imageio/imageio/issues/933

This PR is a collection of optimizations and tweaks to improve the performance of our pyav plugin. It adds the following improvements:

- It removes an explicit (and redundant) colorspace conversion that ffmpeg can handle more efficiently internally (#926)
- It allows floats to be used for `fps`, which allows non-integer framerates (#929).
- It checks the container's `variable_fps` flag when reading metadata to automatically set `constant_framerate` if not manually specified
- It preallocates the output array when bulk reading (`index=...`) if the output shape is known in advance. (#932)
- It optimizes sequential calls to `read` so that they reuse the current decoder state if possible. (#933)
- It exposes the frame type as metadata
- It allows setting `max_keyframe_interval` and `force_keyframes` (open/closed GOP) when writing to control intra frame behavior.

### CODE REVIEW

**Comment by FirefoxMetzger** [comment]
@hmaarrfk This PR implements a lot of your suggestions. Are you interested in reviewing it?

File: imageio/plugins/pyav.py [hunk 1]
```diff
@@ -1047,30 +1080,63 @@ def _unpack_frame(self, frame: av.VideoFrame, *, format: str = None) -> np.ndarr
...
+        self._next_keyframe_idx = idx + self._next_idx
+
     def _seek(self, index, *, constant_framerate: bool = True) -> Generator:
         """Seeks to the frame at the given index."""
 
-        # this may be made faster for formats that have some kind
-        # of keyframe table in their header data.
-        if not constant_framerate:
+        if index == self._next_idx:
+            return  # fast path :)
```
Comment (hmaarrfk) [inline imageio/plugins/pyav.py:1102]: nice way to do it.

**Comment by hmaarrfk** [comment]
I've also found that OpenCV is quite aggressive on the number of threads it uses. Which is why it can seem like opencv is fast.

When i use the same number of threads between OpenCV and pyav, i can close the gap in performance quite significantly.

PyAV still requires a "copy" of the data out of the plantes RGB format, we can make those PRs to PyAV. But I want their help resolving https://github.com/PyAV-Org/PyAV/issues/1045 which is quite important in future workflows i have in mind.

**Comment by FirefoxMetzger** [comment]
almost there :) I have to fix a CI issue (GH actions now uses MacOS 12 which doesn't have pre-built wheels for OpenCV, so we need to exclude that plugin on this MacOS version) and then we can merge.

**Comment by almarklein** [comment]
🚀

**Comment by hmaarrfk** [comment]
Great job! Thanks!

**Pull request merged**